library/file.lisp

```diff
@@ -0,0 +1,724 @@
+
+;;;
+;;; Handling existential path queries
+;;;
+
+(coalton-toplevel
+
+  (declare directory-pathname? ((Into :a Pathname) => :a -> Boolean))
+  (define (directory-pathname? path)
+    "Returns True if a pathname has no file component."
+    (let p = (the Pathname (into path)))
+    (lisp Boolean (p)
+      (to-boolean (uiop:directory-pathname-p p))))
+
+  (declare file-pathname? ((Into :a Pathname) => :a -> Boolean))
+  (define (file-pathname? path)
+    "Returns True if a pathname has a file component."
+    (let p = (the Pathname (into path)))
+    (lisp Boolean (p)
+      (to-boolean (uiop:file-pathname-p p))))
+
+  (declare exists? ((Into :a Pathname) => :a -> (Result FileError Boolean)))
+  (define (exists? path)
+    "Returns whether a file or directory exists."
+    (do
+     (let p = (the Pathname (into path)))
+     (lisp (Result FileError Boolean) (p)
+       (%handle-file-function (to-boolean (cl:probe-file p))))))
+
+  (declare %if-directory-path ((Into :a Pathname)
+                               => (Pathname -> (Result FileError :b))
+                               -> :a
+                               -> (Result FileError :b)))
+  (define (%if-directory-path action path)
+    "Performs an operation only if the path is a valid directory pathname."
+    (do
+     (let p = (the Pathname (Into path)))
+     (let dir = (directory-pathname? p))
+      (if dir
+          (action p)
+          (Err (PathError "Invalid directory path." p)))))
+  
+  (declare directory-exists? ((Into :a Pathname) => :a -> (Result FileError Boolean)))
+  (define (directory-exists? path)
+    "Returns True if a pathname names a directory that exists."
+    (%if-directory-path (fn (p)
+                               (lisp (Result FileError Boolean) (p)
+                                 (%handle-file-function (to-boolean (uiop:directory-exists-p p)))))
+                        path))
+
+  (declare file-exists? ((Into :a Pathname) => :a -> (Result FileError Boolean)))
+  (define (file-exists? path)
+    "Returns True if a pathname names a file that exists."
+    (do
+     (let p = (the Pathname (Into path)))
+     (let file = (file-pathname? p))
+      (if file
+          (lisp (Result FileError Boolean) (p)
+            (%handle-file-function (to-boolean (uiop:file-exists-p p))))
+          (Err (PathError "Invalid file path." p))))))
+
+;;;
+;;; Merging, semigroup and monoid
+;;;
+
+(coalton-toplevel
+
+  (declare merge ((Into :a Pathname) (Into :b Pathname) => :a -> :b -> Pathname))
+  (define (merge path1 path2)
+    "Merges two pathnames together. The directory pathname should be the first argument."
+    (let p1 = (the Pathname (into path1)))
+    (let p2 = (the Pathname (into path2)))
+    (if (not (directory-pathname? p1))
+        (error (PathError "Merge: first argument must be a directory path." p1))
+        (lisp Pathname (p1 p2)
+          (cl:merge-pathnames p2 p1))))
+
+  (define-instance (Semigroup Pathname)
```

Comment:
does this satisfy semigroup laws?

Reply:
I'm not sure, I believe it was required for implementing `monoid` which is necessary for `foldable` etc.